src/databricks/sql/cloudfetch/download_manager.py

```diff
@@ -73,143 +52,49 @@ def get_next_downloaded_file(
         Args:
             next_row_offset (int): The offset of the starting row of the next file we want data from.
         """
-        # No more files to download from this batch of links
-        if not self.download_handlers:
-            self._shutdown_manager()
-            return None
-
-        # Remove handlers we don't need anymore
-        self._remove_past_handlers(next_row_offset)
 
-        # Schedule the downloads
+        # Make sure the download queue is always full
         self._schedule_downloads()
 
-        # Find next file
-        idx = self._find_next_file_index(next_row_offset)
-        if idx is None:
+        # No more files to download from this batch of links
+        if len(self._download_tasks) == 0:
             self._shutdown_manager()
             return None
-        handler = self.download_handlers[idx]
 
-        # Check (and wait) for download status
-        if self._check_if_download_successful(handler):
-            link = handler.result_link
-            logger.debug(
-                "ResultFileDownloadManager: file found for row index {}: start {}, row count: {}".format(
-                    next_row_offset, link.startRowOffset, link.rowCount
-                )
-            )
-            # Buffer should be empty so set buffer to new ArrowQueue with result_file
-            result = DownloadedFile(
-                handler.result_file,
-                handler.result_link.startRowOffset,
-                handler.result_link.rowCount,
-            )
-            self.download_handlers.pop(idx)
-            # Return True upon successful download to continue loop and not force a retry
-            return result
-        else:
+        task = self._download_tasks.pop(0)
+        # Future's `result()` method will wait for the call to complete, and return
+        # the value returned by the call. If the call throws an exception - `result()`
+        # will throw the same exception
+        file = task.result()
+        if (next_row_offset < file.start_row_offset) or (
+            next_row_offset > file.start_row_offset + file.row_count
+        ):
```

Comment:
if `next_row_offset` does not match file's offset, is it intended to return the file?

Reply:
Short answer - yes. The idea is that as user iterates through records from beginning to end, we fetch data sequentially, and fail if next portion of data cannot be fetched. So there are never gaps in the data, and even tracking the current row number should be redundant. Eventually, the whole `next_row_offset` thing should be removed, but for now I keep it and this log for debug purposes

Reply:
I see, but this will violate the API contract. If the client request the file/chunk whose offset is different from current downloaded file, they will get the wrong file. If we want to do this as iterator and does not respect offset, we should remove it from the parameter or create another method if we want to backward compatible?

Reply:
This class is internal to the library, so I wouldn't worry that much. Library's public interface allows only iterating through rows from beginning to end. So user cannot request an arbitrary row. Here, the previous implementation needed a `next_row_offset` to find the correct batch among the downloaded ones, because previous implementation was downloaded batches concurrently and in a random order. But in fact, all the batches already come in a proper order. So if we make sure they're downloaded sequentially (as in new implementation) - there should be no such case when user tries to get next batch and it doesn't match the current row number. And, as a result, `next_row_offset` should not be needed too. I can remove it right now, but I would like to have this log we're duscussing right now - for debug purposes. And after users confirm that new code works properly - I'm going to completely remove `next_row_offset` in the next version

Reply:
If the `next_row_offset` will not be respected and may return the file does not match, I think we should remove it as even if it is used by the client and they call it but get the wrong result, which is also wrong and breaks the contract.

Reply:
@jackyhu-db Yes, I totally agree with you on this. And, as I said, I want to keep it for now for debugging purposes. Currently `next_row_offset` is just passed all the way along from the `ThriftBackend` class and is not used for any other purpose except this log message. After we confirm that users no longer have issues with CloudFetch - I will remove it completely

Reply:
If this is case, can you add a new download manager instead of rewriting existing one and add a flag/option to switch to the new one? Then user can switch it back to the old one if they find some problems.

Reply:
Users can disable CloudFetch feature - that's what they already do when they have issues with this feature. This download manager is for CloudFetch only.

Library does not allow to fetch data from arbitrary place, because Thrift server doesn't support so